File: backend/app/services/text_normalization.py
"""Lightweight multilingual text normalization for user-authored community posts."""

import re
import unicodedata

from app.services.language_utils import detect_language


_ARABIC_CHAR_MAP = str.maketrans(
    {
        "أ": "ا",
        "إ": "ا",
        "آ": "ا",
        "ٱ": "ا",
        "ى": "ي",
        "ؤ": "و",
        "ئ": "ي",
        "ـ": "",
    }
)


def _collapse_spaces(value: str) -> str:
    return re.sub(r"\s+", " ", value).strip()


def normalize_user_text(text: str, *, field: str = "body") -> str:
    value = unicodedata.normalize("NFKC", text or "")
    value = value.replace("_", " ").replace("،", ", ").replace("؛", "; ").replace("?", "? ").replace("؟", "؟ ")
    value = _collapse_spaces(value)
    if not value:
        return ""

    language = detect_language(value)
    if language == "ar":
        value = value.translate(_ARABIC_CHAR_MAP)
        value = _collapse_spaces(value)
        return value

    value = re.sub(r"\s*([,;:.!?])\s*", r"\1 ", value)
    value = _collapse_spaces(value)

    if field == "plant_name":
        return value.title()

    return value[:1].upper() + value[1:] if value else value

File: frontend/lib/text-normalization.ts
function collapseSpaces(value: string) {
  return value.replace(/\s+/g, " ").trim();
}

function detectArabic(value: string) {
  return /[\u0600-\u06FF]/.test(value);
}

export function normalizeUserText(text: string, field: "body" | "plant_name" = "body") {
  let value = (text ?? "").normalize("NFKC");
  value = value.replace(/_/g, " ").replace(/،/g, ", ").replace(/؛/g, "; ").replace(/\?/g, "? ").replace(/؟/g, "؟ ");
  value = collapseSpaces(value);

  if (!value) {
    return "";
  }

  if (detectArabic(value)) {
    value = value
      .replace(/[أإآٱ]/g, "ا")
      .replace(/ى/g, "ي")
      .replace(/ؤ/g, "و")
      .replace(/ئ/g, "ي")
      .replace(/ـ/g, "");
    return collapseSpaces(value);
  }

  value = value.replace(/\s*([,;:.!?])\s*/g, "$1 ");
  value = collapseSpaces(value);

  if (field === "plant_name") {
    return value.replace(/\b\w/g, (char) => char.toUpperCase());
  }

  return value.charAt(0).toUpperCase() + value.slice(1);
}
